src/sqlfluff/core/rules/std/L042.py

```diff
@@ -0,0 +1,64 @@
+"""Implementation of Rule L042."""
+
+from sqlfluff.core.rules.base import BaseCrawler, LintResult
+
+
+class Rule_L042(BaseCrawler):
+    """Join clauses should not contain subqueries. Use CTEs instead.
+
+    NB: Some dialects don't allow CTEs, and for those dialects
+    this rule makes no sense and should be disabled.
+
+    | **Anti-pattern**
+
+    .. code-block:: sql
+
+        select
+            a.x, a.y, b.z
+        from a
+        join (
+            select x, z from b
+        ) using(x)
+
+
+    | **Best practice**
+
+    .. code-block:: sql
+
+        with c as (
+            select x, z from b
+        )
+        select
+            a.x, a.y, c.z
+        from a
+        join c using(x)
+
+    """
+
+    def _eval(self, segment, **kwargs):
+        """Join clauses should not contain subqueries. Use CTEs instead.
+
+        NB: No fix for this routine because it would be very complex to
+        implement reliably.
+        """
+        if segment.is_type("join_clause"):
+            # Get the referenced table segment
+            table_expression = segment.get_child("table_expression")
```

Comment:
Does this handle the case of multiple joins, e.g.:
```
SELECT c
FROM t1
JOIN t2
ON <<something>>
JOIN (
    SELECT x, z FROM t3
)
```

It may very well work, mainly I'm unsure how the grammar is structured. Worth adding a test for this?

Reply:
It should cover it yes, but I'll add a test case to confirm that.